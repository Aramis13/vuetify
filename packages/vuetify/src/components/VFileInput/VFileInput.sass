@import '../../styles/styles.sass'

+theme(v-file-input) using ($material)
  .v-file-input__text
    color: map-deep-get($material, 'text', 'primary')

    &--placeholder
      color: map-deep-get($material, 'text', 'secondary')

// Block
.v-file-input
  input[type="file"]
    opacity: 0
    width: 0

// Element
.v-file-input .v-file-input__text
  display: flex
  align-items: center
  width: 100%

  &.v-file-input__text--chips
    flex-wrap: wrap

  .v-chip
    margin: 4px
