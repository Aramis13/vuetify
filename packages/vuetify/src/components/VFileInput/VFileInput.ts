// Styles
import './VFileInput.sass'

// Extensions
import VTextField from '../VTextField'

// Components
import { VProgressLinear } from '../VProgressLinear'
import { VChip } from '../VChip'

// Types
import { VNode } from 'vue'
import { PropValidator } from 'vue/types/options'

// Helpers
import { wrapInArray, humanReadableFileSize } from '../../util/helpers'

export default VTextField.extend({
  name: 'v-file-input',

  model: {
    prop: 'value',
    event: 'change',
  },

  props: {
    accept: {
      type: String,
      default: null,
    } as PropValidator<string | null>,
    chips: Boolean,
    clearable: {
      type: Boolean,
      default: true,
    },
    counterSizeString: {
      type: String,
      default: '$vuetify.fileInput.counterSize',
    },
    counterString: {
      type: String,
      default: '$vuetify.fileInput.counter',
    },
    deletableChips: {
      type: Boolean,
      default: true,
    },
    displaySize: {
      type: [Boolean, Number],
      default: false,
      validator: v => typeof v === 'boolean' || v === 1000 || v === 1024,
    } as PropValidator<boolean | 1000 | 1024>,
    multiple: Boolean,
    placeholder: String,
    prependIcon: {
      type: String,
      default: '$vuetify.icons.file',
    },
    progress: {
      type: [Number, String],
      default: null,
      validator: (v: number | string) => !isNaN(parseInt(v)),
    } as PropValidator<number | string | null>,
    readonly: {
      type: Boolean,
      default: true,
    },
    smallChips: Boolean,
    type: {
      type: String,
      default: 'file',
    },
    value: {
      type: [Object, Array],
      default: () => ([]),
    } as PropValidator<File | File[]>,
  },

  data () {
    return {
      lazyValue: wrapInArray(this.value) as File[],
    }
  },

  computed: {
    classes (): object {
      return {
        ...VTextField.options.computed.classes.call(this),
        'v-file-input': true,
      }
    },
    counterValue (): string {
      if (this.displaySize) {
        const bytes = this.internalValue.length > 0
          ? (this.internalValue as File[]).map(f => f.size).reduce((a, b) => a + b)
          : 0

        return this.$vuetify.lang.t(this.counterSizeString, this.lazyValue.length, humanReadableFileSize(bytes, this.base))
      }
      return this.$vuetify.lang.t(this.counterString, this.lazyValue.length)
    },
    internalValue: {
      get (): File[] {
        return this.lazyValue
      },
      set (val: any) {
        this.lazyValue = wrapInArray(val)
        this.$emit('change', this.lazyValue)
      },
    },
    isDirty (): boolean {
      return this.internalValue.length > 0
    },
    isLabelActive (): boolean {
      return this.isDirty
    },
    text (): string[] {
      if (!this.isDirty) return [this.placeholder]

      return this.internalValue.map((file: File) =>
        this.displaySize ? `${file.name} (${humanReadableFileSize(file.size, this.base)})` : file.name
      )
    },
    base (): 1000 | 1024 | undefined {
      return typeof this.displaySize !== 'boolean' ? this.displaySize : undefined
    },
    hasChips (): boolean {
      return this.chips || this.smallChips
    },
  },

  methods: {
    clearableCallback () {
      this.internalValue = []
    },
    genPrependSlot () {
      const icon = this.genIcon('prepend', () => {
        this.$refs.input.click()
      })

      icon.data!.attrs! = { tabindex: 0 }

      return this.genSlot('prepend', 'outer', [icon])
    },
    genInput () {
      const input = VTextField.options.methods.genInput.call(this)

      input.data!.attrs!.multiple = this.multiple
      input.data!.attrs!.accept = this.accept
      delete input.data!.domProps!.value

      return [
        this.genText(),
        input,
      ]
    },
    genText () {
      const children = this.$scopedSlots.selection
        ? this.$scopedSlots.selection({ text: this.text, files: this.internalValue })
        : this.hasChips && this.isDirty ? this.genChips() : [this.text.join(', ')]

      return this.$createElement('div', {
        staticClass: 'v-file-input__text',
        class: {
          'v-file-input__text--placeholder': this.placeholder && !this.isDirty,
          'v-file-input__text--chips': this.hasChips && !this.$scopedSlots.selection,
        },
        on: {
          click: () => {
            this.$refs.input.click()
          },
        },
      }, children)
    },
    genChips () {
      if (!this.isDirty) return []

      return this.text.map((text, i) => this.$createElement(VChip, {
        props: {
          close: this.deletableChips,
          small: this.smallChips,
        },
        on: {
          'click:close': () => {
            this.internalValue = this.internalValue.filter((val: File[], index: number) => {
              return i !== index
            })
          },
        },
      }, [text]))
    },
    genProgress (): VNode | VNode[] | null {
      if (this.loading === false && !this.progress) return null

      return this.$slots.progress || this.$createElement(VProgressLinear, {
        props: {
          absolute: true,
          color: (this.loading === true)
            ? (this.color || 'primary')
            : (this.loading || 'primary'),
          height: this.loaderHeight,
          indeterminate: !this.progress,
          value: this.progress || 0,
        },
      })
    },
    onInput (e: Event) {
      const target = e.target as HTMLInputElement
      this.internalValue = [...target.files]
    },
  },
})
