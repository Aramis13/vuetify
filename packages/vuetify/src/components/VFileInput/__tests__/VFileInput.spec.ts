// Components
import VFileInput from '../VFileInput'

// Services
import { Lang } from '../../../services/lang'

// Libraries
import {
  Wrapper,
  mount,
  MountOptions,
} from '@vue/test-utils'

const oneMBFile = { name: 'test', size: 1048576 }
const twoMBFile = { name: 'test', size: 2097152 }

describe('VFileInput.ts', () => {
  type Instance = InstanceType<typeof VFileInput>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(VFileInput, {
        mocks: {
          $vuetify: {
            lang: new Lang(),
          },
        },
        ...options,
      })
    }
  })

  it('should render', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render counter', () => {
    const wrapper = mountFunction({
      propsData: {
        counter: true,
      },
    })

    wrapper.setData({
      lazyFileValue: [ oneMBFile ],
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should display file size', () => {
    const wrapper = mountFunction({
      propsData: {
        displaySize: true,
      },
    })

    wrapper.setData({
      lazyFileValue: [ twoMBFile ],
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      displaySize: 1000,
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should display total size in counter', () => {
    const wrapper = mountFunction({
      propsData: {
        displaySize: true,
        counter: true,
      },
    })

    wrapper.setData({
      lazyFileValue: [ oneMBFile, twoMBFile ],
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      displaySize: 1000,
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be unclearable', () => {
    const wrapper = mountFunction({
      propsData: {
        clearable: false,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should emit event on change', () => {
    const fn = jest.fn()
    const wrapper = mountFunction({
      listeners: {
        'update:fileValue': fn,
      },
    })

    wrapper.setData({
      lazyFileValue: [ oneMBFile ],
    })

    expect(fn).toHaveBeenLastCalledWith([ oneMBFile ])
  })

  it('should display progress', () => {
    const wrapper = mountFunction({
      propsData: {
        progress: 30,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      progress: '70',
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should work with accept', () => {
    const wrapper = mountFunction({
      propsData: {
        accept: 'image/*',
      },
    })

    expect(wrapper.find('input').element.getAttribute('accept')).toBe('image/*')
  })

  it('should disable file input', () => {
    const wrapper = mountFunction({
      propsData: {
        disabled: true,
      },
    })

    expect(wrapper.find('input').element.getAttribute('disabled')).toBe('disabled')
  })
})
