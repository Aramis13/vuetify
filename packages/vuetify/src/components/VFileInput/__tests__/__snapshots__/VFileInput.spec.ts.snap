// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VFileInput.ts should be unclearable 1`] = `

<div class="v-input v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
        </div>
        <input id="input-25"
               readonly="readonly"
               type="file"
        >
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should display file size 1`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
          test (2.0 MiB)
        </div>
        <input id="input-13"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should display file size 2`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
          test (2.1 MB)
        </div>
        <input id="input-13"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should display progress 1`] = `

<div class="v-input v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
        </div>
        <input id="input-30"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
          </i>
        </div>
      </div>
      <div role="progressbar"
           aria-valuemin="0"
           aria-valuemax="100"
           aria-valuenow="30"
           class="v-progress-linear v-progress-linear--absolute theme--light"
           style="height: 2px;"
      >
        <div class="v-progress-linear__background primary"
             style="opacity: 0.3; width: 100%;"
        >
        </div>
        <div class="v-progress-linear__buffer">
        </div>
        <div class="v-progress-linear__determinate primary"
             style="width: 30%;"
        >
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should display progress 2`] = `

<div class="v-input v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
        </div>
        <input id="input-30"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
          </i>
        </div>
      </div>
      <div role="progressbar"
           aria-valuemin="0"
           aria-valuemax="100"
           aria-valuenow="70"
           class="v-progress-linear v-progress-linear--absolute theme--light"
           style="height: 2px;"
      >
        <div class="v-progress-linear__background primary"
             style="opacity: 0.3; width: 100%;"
        >
        </div>
        <div class="v-progress-linear__buffer">
        </div>
        <div class="v-progress-linear__determinate primary"
             style="width: 70%;"
        >
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should display total size in counter 1`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
          test (1.0 MiB), test (2.0 MiB)
        </div>
        <input id="input-19"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
      <div class="v-counter theme--light">
        2 files (3.0 MiB in total)
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should display total size in counter 2`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
          test (1.0 MB), test (2.1 MB)
        </div>
        <input id="input-19"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
      <div class="v-counter theme--light">
        2 files (3.1 MB in total)
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should render 1`] = `

<div class="v-input v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
        </div>
        <input id="input-1"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should render chips 1`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text v-file-input__text--chips">
          <span class="v-chip v-chip--no-color v-chip--removable theme--light v-size--default">
            <span class="v-chip__content">
              test
              <i role="button"
                 class="v-icon notranslate v-chip__close v-icon--link v-icon--right material-icons theme--light"
              >
                $vuetify.icons.delete
              </i>
            </span>
          </span>
        </div>
        <input id="input-70"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should render counter 1`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text">
          test
        </div>
        <input id="input-7"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
      <div class="v-counter theme--light">
        1 files
      </div>
    </div>
  </div>
</div>

`;

exports[`VFileInput.ts should render small chips 1`] = `

<div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field v-file-input">
  <div class="v-input__prepend-outer">
    <div tabindex="0"
         class="v-input__icon v-input__icon--prepend"
    >
      <i role="button"
         class="v-icon notranslate v-icon--link material-icons theme--light"
      >
        $vuetify.icons.file
      </i>
    </div>
  </div>
  <div class="v-input__control">
    <div class="v-input__slot">
      <div class="v-text-field__slot">
        <div class="v-file-input__text v-file-input__text--chips">
          <span class="v-chip v-chip--no-color v-chip--removable theme--light v-size--small">
            <span class="v-chip__content">
              test
              <i role="button"
                 class="v-icon notranslate v-chip__close v-icon--link v-icon--right material-icons theme--light"
              >
                $vuetify.icons.delete
              </i>
            </span>
          </span>
        </div>
        <input id="input-78"
               readonly="readonly"
               type="file"
        >
      </div>
      <div class="v-input__append-inner">
        <div class="v-input__icon v-input__icon--clear">
          <i role="button"
             class="v-icon notranslate v-icon--link material-icons theme--light"
          >
            $vuetify.icons.clear
          </i>
        </div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light">
        <span name="message-transition"
              tag="div"
              class="v-messages__wrapper"
        >
        </span>
      </div>
    </div>
  </div>
</div>

`;
